src/binary_format.cc

```diff
@@ -101,67 +82,53 @@ void FactorGraph::load_variables(const std::string &filename) {
             << std::endl;
         std::abort();
     }
-    bool is_evidence = isevidence >= 1;
-    bool is_observation = isevidence == 2;
+    bool is_evidence = role_serialized >=
+                       1;  // TODO interpret as bit vector instead of number?
+    bool is_observation = role_serialized == 2;
     variable_value_t init_value = is_evidence ? initial_value : 0;
-
-    variables[id] = RawVariable(id, type_const, is_evidence, cardinality,
-                                init_value, init_value, is_observation);
-
+    variables[vid] = RawVariable(vid, dtype, is_evidence, cardinality,
+                                 init_value, init_value, is_observation);
     ++size.num_variables;
     if (is_evidence) {
       ++size.num_variables_evidence;
     } else {
       ++size.num_variables_query;
     }
   }
-  file.close();
 }
 
 void FactorGraph::load_factors(const std::string &filename) {
-  std::ifstream file;
-  file.open(filename.c_str(), std::ios::in | std::ios::binary);
-  variable_id_t variable_id;
-  weight_id_t weightid;
-  variable_value_t value_id;
-  int16_t type;
-  factor_arity_t arity;
-  variable_value_t should_equal_to;
+  std::ifstream file(filename, std::ios::in | std::ios::binary);
   while (file.good()) {
-    file.read((char *)&type, 2);
-    if (!file.read((char *)&arity, 4)) break;
-
-    type = be16toh(type);
-    arity = be32toh(arity);
-
+    uint16_t type;
+    factor_arity_t arity;
+    // read fields
+    read_be(file, type);
+    if (!read_be(file, arity)) break;
```

Comment:
Honestly, I couldn't understand either.  I tried to put `assert` around other lines, but that crashes normal cases.  Maybe time to search for the right idiom for using `ifstream` with `ios::binary`?  Will add some comments or change into a clearer flow once I figure out.